finality-aleph/src/party.rs

```diff
@@ -431,25 +431,32 @@ where
                 }
             }
         };
-        let session = AuthoritySession {
-            session_id,
-            stop_h: last_block_of_session::<B>(session_id, self.session_period),
-            authorities,
-        };
-
-        self.sessions.lock().insert(session_id, session.clone());
+        self.session_authorities
+            .lock()
+            .insert(session_id, authorities.clone());
+        let last_block = last_block_of_session::<B>(session_id, self.session_period);
+
+        // Early skip attempt -- this will trigger during catching up (initial sync).
+        if self.client.info().best_number >= last_block {
+            // We need to give the JustificationHandler some time to pick up the keybox for the new session,
+            // validate justifications and finalize blocks.
+            Delay::new(Duration::from_millis(2000)).await;
+            let last_finalized_number = self.client.info().finalized_number;
+            if last_finalized_number >= last_block {
```

Comment:
Actually, could you do this check both before and after the delay? Otherwise we might be waiting for completely no reason.

Reply:
Right, I was surprised you didn't comment on that :) Yeah, this is bad, I changed it now into a loop that checks 10 times, every 200ms. It's a little ugly, but works much better.